### ISSUE TITLE
[HOLD for payment 2024-07-10] [HOLD for payment 2024-07-02] LOW: [$500] Stop using `reportAction.originalMessage` or `reportAction.message.text`

### ISSUE THREAD

**Issue by quinthar** [Expensify/App#39797]
Problem:
====
Each comment in the server database is stored in `reportAction.message` in HTML form.  However, for some reason lost to time, when we send that same comment out as an Onyx update, we:

1. Rename the original message in the database to `reportAction.originalMessage` in the update (which is confusing)
2. Create a new `message` object that contains a copy of the `html` message, as well as a stripped `text` version.
3. Put the `message` object inside a totally unnecessary one-element array

That means every comment is actually sent out *three times* (four, if you include the update to the `report` object, but that's out of scope for now).  This is clearly wasteful in a number of ways:

1. It requires postprocessing on the server to do this extraneous operation
2. It increases network bandwidth by 3x
3. It increases RAM use 3x
4. It increases on-device storage 3x
5. It's confusing AF

Additionally, it's _particularly_ problematic when sending Onyx updates out via UrbanAirship, which is extremely limited in the payload sizes allowed -- with anything over a certain limit just dropped quietly and never delivered, causing "gaps" in our update stream (which require more network calls to "backfill" the Onyx data on app open, which makes things slow).  Basically, this was just a mistake introduced for reasons we can't remember, but that we want to undo.

Solution:
====
To solve this, please:

1. **Phase 1 (External / clientside):** Stop using `originalMessage`, start using array-`message`
    1. Update every instance of `originalMessage.html` to use `message.html || message[0].html`
    2. Update every instance of `message.text` to dynamically strip the text from `message.html || message[0].html`
    3. Update every key within originalMessage to use the original message
2. **Phase 2 (Internal / serverside):** Once we've confirmed the client no longer uses `originalMessage` or `message.text` or any other key within originalMessage:
    1. Stop sending `originalMessage`
    2. Start sending `message` without an array
4. **Phase 3 (External / clientside):** Remove checking for "array message"

This issue is for phase 1 and phase 3.

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01856d1399295ef760</li>
        <li>Upwork Job ID: 1776729925881405440</li>
        <li>Last Price Increase: 2024-07-12</li>
<li>Automatic offers: </li>
<ul>
<li>cubuspl42 | Reviewer | 0</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @puneetlath</details>

**Comment by maxim-grin** [comment]
Hey I’m going to investigate why the payload is so big in Onyx storage. Possibly track what change exactly caused the problem. Then either revert that piece of change or use one of suggestions by @thienlnam

**Comment by maxim-grin** [comment]
Hey I’m going to investigate why the payload is so big in Onyx storage. Possibly track what change exactly caused the problem. Then either revert that piece of change or use one of suggestions by @thienlnam 

Contributor details
Your Expensify account email: maxim.grin@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01fb7b172c01252917

**Comment by cubuspl42** [comment]
@ShridharGoel 

> Similarly, check all usages and update them

Are you saying to literally repeat this code snippet in all places? Couldn't we refactor the code to avoid this repetition?

As the scope of the issue is well-defined and rather narrow, I believe this part of the plan can be a part of the proposal.

@nkdengineer 

Aren't you quoting browser-specific local notifications code? 🤔 As I understand it, the ultimate goal is to optimize the payload size of remote push notifications.

**Comment by nkdengineer** [comment]
> Aren't you quoting browser-specific local notifications code? 🤔 As I understand it, the ultimate goal is to optimize the payload size of remote push notifications.

@cubuspl42 In the OP it's mentioned
> @thienlnam thinks we added `text` for notifications

So I'm quoting the place that we're showing notification based on `text`.

It's only 1 part of the proposal though, what do you think about the rest of my proposal?

> Are you saying to literally repeat this code snippet in all places? Couldn't we refactor the code to avoid this repetition?

@cubuspl42 My alternative solution achieves this

**Comment by cubuspl42** [comment]
@quinthar

Would you summarize the Expensify notifications architecture in a few sentences so that we can be more confident with proposal preparation and review?

As I understand that most (if not all?) notification processing on mobile Native is server-side (push-based), and we're talking about optimizing the pushed payload size. But I'm not 100% confident how that relates to the Onyx schema.

Also, there were some doubts about the scope...

> I'm assuming here that we're only omitting the text from reportActions in urban airship native updates, if we're going to remove text in Pusher update too, then we need to populate the text in the Pusher report actions updates by htmlToText.

Would you clarify this too? My intuition was that we wanted to nuke all usage of the `text` property client-side, also removing it from the typing. But it would be good to be sure about it.

**Comment by quinthar** [comment]
Updated the proposal again (sorry!!!) in light of `message` being in this weird one-element array.  Thanks for your understanding and help in cleaning up this mess!

**Comment by nkdengineer** [comment]
[Proposal updated](https://github.com/Expensify/App/issues/39797#issuecomment-2042472371) after latest requirements update

**Comment by cubuspl42** [comment]
@nkdengineer I think your proposal shows more preparation. How is your capacity, though? Would you be able to start working on this within 24-48 hours?

**Comment by nkdengineer** [comment]
> Would you be able to start working on this within 24-48 hours?

@cubuspl42 Yes, I definitely can

**Comment by nkdengineer** [comment]
@thienlnam Hi! I see this is a `HIGH` priority issue so it should start at $500 (from the announcement [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1712256409152489))

> While bugs/feature requests will now start at $250, any that are deemed Critical or High to our roadmap (noted via the label `High Priority`) will automatically be updated to $500. This is to ensure we get the necessary attention and urgency applied to these issues.

Perhaps the automation did not work well in this case.

If that's right, could you help to adjust the bounty here, thanks 🙏

**Comment by cubuspl42** [comment]
Hm, it's likely because the issue is missing the "High priority" label. But I'm not sure if the "HIGH:" prefix and the "High priority" label represent the same concept 🤔

**Comment by nkdengineer** [comment]
We already have a draft PR [here](https://github.com/Expensify/App/pull/40168) to fix things in phase 1, @thienlnam please let me know when phase 2 is completed on the server side, I will then update phase 3 accordingly.
Cc @cubuspl42

**Comment by quinthar** [comment]
Great work!  What's your ETA for submitting the PR for review?

**Comment by nkdengineer** [comment]
I will open PR tomorrow

**Comment by cubuspl42** [comment]
@nkdengineer Just so we're on the same page, phase 1 must be deployed to production first. Also, some time has to pass, so Native users have the time to update. So we're not blocked by phase 2 in any way.

**Comment by nkdengineer** [comment]
@cubuspl42 The PR is [here](https://github.com/Expensify/App/pull/40168).

**Comment by quinthar** [comment]
Great!  @cubuspl42 what is your ETA for review?

**Comment by cubuspl42** [comment]
I'm on it 🙂

**Comment by quinthar** [comment]
Great, what is your ETA for review?

**Comment by cubuspl42** [comment]
It was yesterday. I [raised concerns](https://github.com/Expensify/App/pull/40168#issuecomment-2060947228) about whether we actually covered the whole scope of the Phase 1. I even pinged you 🙂 We can continue the dialogue on the PR.

**Comment by mallenexpensify** [comment]
@parasharrajat reassigning, please take over as C+. If you don't have bandwidth, unassign yourself. Thanks.  PR is
- https://github.com/Expensify/App/pull/40168

**Comment by mallenexpensify** [comment]
Thanks @parasharrajat , 👋 @c3024 
@c3024 [commented on the PR ](https://github.com/Expensify/App/pull/40168#issuecomment-2084705409)earlier today

**Comment by c3024** [comment]
@thienlnam @quinthar 

The issue talks about only `text` and `html`

> Phase 1 (External / clientside): Stop using originalMessage, start using array-message
Update every instance of originalMessage.html to use message.html || message[0].html
Update every instance of message.text to dynamically strip the text from message.html || message[0].html

but `originalMessage` has many more fields like `reason`, `resolution`, `IOUTransactionID`, `events`, `oldName`, `newName`, `participantAccountIDs`, etc. which are not present in `message`. So this phase 1 should be to update every instance of `originalMessage` (not just fields of `html` or `text`) with array `message`

Also in phase 2

>Phase 2 (Internal / serverside): Once we've confirmed the client no longer uses originalMessage or message.text:
Stop sending originalMessage
Start sending message without an array

`message` should be sent including all the above mentioned other fields to completely deprecate `originalMessage`.

Can you please update phase 1 and 2 of the [issue description](https://github.com/Expensify/App/issues/39797#issue-2229436084) to include these?

**Comment by mallenexpensify** [comment]
Added `Bug` to assign a BZ. @quinthar @thienlnam , can you address @c3024 's request above when you have a min
https://github.com/Expensify/App/issues/39797#issuecomment-2130753109

**Comment by thienlnam** [comment]
I've updated the description

**Comment by roryabraham** [comment]
I did some exploratory work on this topic in https://github.com/Expensify/App/pull/41872 - maybe useful for you @nkdengineer @c3024 @thienlnam

**Comment by roryabraham** [comment]
self-assigning as I've been pulled into the PR for review

**Comment by puneetlath** [comment]
@nkdengineer @c3024 how is this going? Are we almost done?

**Comment by muttmuure** [comment]
Is this being worked on?

**Comment by thienlnam** [comment]
Still in review - pending a review from @roryabraham. Dropping my assignment since I'm not adding much value as a reviewer here

**Comment by c3024** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
 
* [x]  [@c3024] The PR that introduced the bug has been identified. Link to the PR: This is an improvement and not a bug.
* [x]  [@c3024] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: NA
* [x]  [@c3024] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: No since this is not a bug.
* [x]  [@c3024] Determine if we should create a regression test for this bug. No, this touches several parts of the repo. There are individual regression tests for individual cases.
* [x]  [@c3024] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. NA

**Comment by c3024** [comment]
>bump on the checklist so we can pay you. 

Done

>Can you also link me your Upwork profile so I can send you a job?

Here it is - https://www.upwork.com/freelancers/~0105555e2f227dbf47

Can the pay be bumped up here? This is a large PR touching many files and there were some big revisions as well. Thanks!

**Comment by puneetlath** [comment]
@nkdengineer please accept offer here: https://www.upwork.com/nx/wm/offer/103096930
@c3024 please accept offer here: https://www.upwork.com/nx/wm/offer/103096934

Both of you, please ping me on this issue when you've accepted.

**Comment by c3024** [comment]
@puneetlath thanks, accepted!

**Comment by puneetlath** [comment]
@c3024 has been paid. Just waiting for you now @nkdengineer.

**Comment by nkdengineer** [comment]
@puneetlath Accepted thanks 🙇

**Comment by puneetlath** [comment]
All paid. Thanks everyone!
